src/RustyObjectStore.jl

```diff
@@ -82,17 +82,60 @@ const DEFAULT_CONFIG = StaticConfig(
     concurrency_limit=512
 )
 
+function default_panic_hook()
+    println("Rust thread panicked, exiting the process")
+    exit(1)
+end
+
 const _OBJECT_STORE_STARTED = Ref(false)
 const _INIT_LOCK::ReentrantLock = ReentrantLock()
+_PANIC_HOOK::Function = default_panic_hook
 
 struct InitException <: Exception
     msg::String
     return_code::Cint
 end
 
-function default_panic_hook()
-    println("Rust thread panicked, exiting the process")
-    exit(1)
+Base.@ccallable function panic_hook_wrapper()::Cint
+    global _PANIC_HOOK
+    _PANIC_HOOK()
+    return 0
+end
+
+# This is the callback that Rust calls to notify a Julia task of a completed operation.
+# The argument is transparent to Rust and is simply what gets passed from Julia in the handle
+# argument of the @ccall. Currently we pass a pointer to a Base.Event that must be notified to
+# wakeup the appropriate task.
+Base.@ccallable function notify_result(event_ptr::Ptr{Nothing})::Cint
+    event = unsafe_pointer_to_objref(event_ptr)::Base.Event
+    notify(event)
+    return 0
+end
+
+# A dict of all tasks that are waiting some result from Rust
+# and should thus not be garbage collected.
+# This copies the behavior of Base.preserve_handle.
+const tasks_in_flight = IdDict{Task, Int64}()
+const preserve_task_lock = Threads.SpinLock()
```

Comment:
Last comment: if we're still seeing high contention here, it could make sense to shard this into multiple dicts and locks, keyed by a hash of the task's objectid? That would be a simple way to significantly reduce contention on the spinlock. Ideally, we'd be able to register a new IO without any spinning at all. I think even just 10 or 100 shards should be enough to completely eliminate that, yeah?